test/extensions/filters/http/adaptive_concurrency/concurrency_controller/gradient_controller_test.cc

```diff
@@ -0,0 +1,464 @@
+    seconds: 30
+  request_count: 5
+)EOF";
+
+  auto controller = makeController(yaml);
+  EXPECT_EQ(controller->concurrencyLimit(), 1);
+
+  // Get minRTT measurement out of the way.
+  advancePastMinRTTStage(controller, yaml, std::chrono::milliseconds(5));
+
+  // Force the limit calculation to run a few times from some measurements.
+  for (int sample_iters = 0; sample_iters < 5; ++sample_iters) {
+    const auto last_concurrency = controller->concurrencyLimit();
+    for (int ii = 1; ii <= 5; ++ii) {
+      tryForward(controller, true);
+      controller->recordLatencySample(std::chrono::milliseconds(4));
+    }
+    time_system_.sleep(std::chrono::milliseconds(101));
+    dispatcher_->run(Event::Dispatcher::RunType::Block);
+    // Verify the value is growing.
+    EXPECT_GT(controller->concurrencyLimit(), last_concurrency);
+  }
+
+  const auto rq_outstanding = [this]() {
+    return stats_.gauge("test_prefix.rq_outstanding", Stats::Gauge::ImportMode::Accumulate).value();
+  };
+
+  // Verify the outstanding requests make sense.
+  for (int ii = 0; ii < controller->concurrencyLimit(); ++ii) {
+    EXPECT_EQ(ii, rq_outstanding());
+    tryForward(controller, true);
+    EXPECT_EQ(ii + 1, rq_outstanding());
+  }
+
+  // The outstanding requests should never exceed the concurrency limit.
+  tryForward(controller, false);
+  EXPECT_EQ(controller->concurrencyLimit(), rq_outstanding());
+  tryForward(controller, false);
+  EXPECT_EQ(controller->concurrencyLimit(), rq_outstanding());
+}
+
+TEST_F(GradientControllerTest, MinRTTRescheduleTest) {
+  const std::string yaml = R"EOF(
+sample_aggregate_percentile: 0.5
+concurrency_limit_params:
+  max_gradient: 3.0
+  max_concurrency_limit: 
+  concurrency_update_interval: 
+    nanos: 100000000 # 100ms
+min_rtt_calc_params:
+  interval:
+    seconds: 30
+  request_count: 5
+)EOF";
+
+  auto controller = makeController(yaml);
+  EXPECT_EQ(controller->concurrencyLimit(), 1);
+
+  // Get initial minRTT measurement out of the way.
+  advancePastMinRTTStage(controller, yaml, std::chrono::milliseconds(5));
+
+  // Force the limit calculation to run a few times from some measurements.
+  for (int sample_iters = 0; sample_iters < 5; ++sample_iters) {
+    const auto last_concurrency = controller->concurrencyLimit();
+    for (int ii = 1; ii <= 5; ++ii) {
+      tryForward(controller, true);
+      controller->recordLatencySample(std::chrono::milliseconds(4));
+    }
+    time_system_.sleep(std::chrono::milliseconds(101));
+    dispatcher_->run(Event::Dispatcher::RunType::Block);
+    // Verify the value is growing.
+    EXPECT_GT(controller->concurrencyLimit(), last_concurrency);
+  }
+
+  // Wait until the minRTT recalculation is triggered again and verify the limit drops.
+  time_system_.sleep(std::chrono::seconds(31));
+  dispatcher_->run(Event::Dispatcher::RunType::Block);
+  EXPECT_EQ(controller->concurrencyLimit(), 1);
+
+  // Verify sample recalculation doesn't occur during the minRTT window.
```

Comment:
This test actually makes me think about what happens after the minRTT window, where we push the limit to 1. Once the window is excited we wait for the next sample window to increase the limit. And we increase from 1.

1. Am I wrong about that analysis?
2. Should we be starting from 1 every time after minRTT probing if 1 is true?